File: 7apps/code/WebApp/web_02_03_better_labels_with_templates/app/scripts/view_controller.js
/***
 * Excerpted from "Seven Mobile Apps in Seven Weeks",
 * published by The Pragmatic Bookshelf.
 * Copyrights apply to this code. It may not be used to create training material,
 * courses, books, articles, and the like. Contact us if you are in doubt.
 * We make no guarantees that this code is fit for any purpose.
 * Visit http://www.pragmaticprogrammer.com/titles/7apps for more book information.
***/
(function($) {

  var namespaces = $.app.namespaces,
      clock = namespaces.models.Clock,
      timeZoneManager = namespaces.managers.TimeZoneManager,
      clockList = $("#clockList"),
      zoneList = $("#zoneList"),
      addClockLink = $("a#addClockLink");

  var MainViewController = {


    initialize: function() {
      this.openZoneListFunction = _.bind(this.addClockClicked, this);
      this.closeZoneListFunction = _.bind(this.dismissZoneList, this);
      addClockLink.click(this.openZoneListFunction);
      zoneList.hide();
      this.refreshClockList();
      clock.start();

      timeZoneManager.fetchTimeZones();
    },

    addClockClicked : function() {
      if (zoneList.children().length === 0) {
        var zones = timeZoneManager.allZones(),
            clickHandler = _.bind(this.zoneClicked, this),
            template = $("#timeZoneTemplate").text();
        _.each(zones, function(zone, index) {
          var item = $(Mustache.render(template, zone));
          item.data("zoneIndex", index);
          item.click(clickHandler);
          zoneList.append(item);
        });
      }
      this.presentZoneList();
    },

    zoneClicked : function(event) {
      var item = $(event.currentTarget),
          index = item.data("zoneIndex");
      timeZoneManager.saveZoneAtIndex(index);
      this.dismissZoneList();
      this.refreshClockList();
    },

    presentZoneList : function() {
      addClockLink.text("Cancel");
      addClockLink.off("click");
      addClockLink.click(this.closeZoneListFunction);
      zoneList.show();
    },

    dismissZoneList : function() {
      addClockLink.text("Add Clock");
      addClockLink.off("click");
      addClockLink.click(this.openZoneListFunction);
      zoneList.hide();
    },

    refreshClockList : function() {
      var zones = timeZoneManager.savedZones(true),
          template = $("#clockTemplate").text();
      clockList.empty();
      _.each(zones, function(zone, index) {
        this.createClock(zone, index, template);
      }, this);
      clock.tick();
    },

    createClock : function(zone, index, template) {
      var item = $(Mustache.render(template, zone));
      clockList.append(item);
    }

  };

  $.app.register("controllers.MainViewController", MainViewController);

})(jQuery);

File: 7apps/code/Xamarin/xamarin_03_01_form_project/Calc/iOS/ViewController.cs
/***
 * Excerpted from "Seven Mobile Apps in Seven Weeks",
 * published by The Pragmatic Bookshelf.
 * Copyrights apply to this code. It may not be used to create training material,
 * courses, books, articles, and the like. Contact us if you are in doubt.
 * We make no guarantees that this code is fit for any purpose.
 * Visit http://www.pragmaticprogrammer.com/titles/7apps for more book information.
***/
﻿using System;
		
using UIKit;
using Foundation;
using System.Drawing;
using CoreGraphics;

namespace Calc.iOS {
  public partial class ViewController : UIViewController, IUICollectionViewDataSource, IUICollectionViewDelegate {

    CalculatorModel model;

		public ViewController(IntPtr handle) : base (handle) {
      model = new CalculatorModel();
      model.OnDisplayChanged += delegate (CalculatorModel model) {
        DisplayLabel.Text = model.Display;
      };
    }

		public override void ViewDidLoad() {
			base.ViewDidLoad();
      DisplayLabel.AccessibilityIdentifier = "calculator_display";
      ButtonCollectionView.RegisterClassForCell(typeof(ButtonCell), ButtonCell.cellId);
      ButtonCollectionView.Delegate = this;
      ButtonCollectionView.DataSource = this;
		}

    public override void ViewDidAppear(bool animated) {
      base.ViewDidAppear(animated);
      ConfigureStyles();
      ButtonCollectionView.ReloadData();
    }

    public override void ViewWillTransitionToSize(CGSize toSize, IUIViewControllerTransitionCoordinator coordinator) {
      base.ViewWillTransitionToSize(toSize, coordinator);
      coordinator.AnimateAlongsideTransition(
        (context) => { ButtonCollectionView.ReloadData(); },
        completion: null
      );
    }

    public nint GetItemsCount(UICollectionView collectionView, nint section) {
      return CalculatorModel.buttons.Length;
    }

    public UICollectionViewCell GetCell(UICollectionView collectionView, NSIndexPath indexPath) {
      var buttonCell = (ButtonCell)collectionView.DequeueReusableCell(ButtonCell.cellId, indexPath);
      var button = CalculatorModel.buttons[indexPath.Row];
      buttonCell.SetTitle(button);
      return buttonCell;
    }

    [Export("collectionView:didSelectItemAtIndexPath:")]
    public void ItemSelected(UICollectionView collectionView, NSIndexPath indexPath) {
      collectionView.DeselectItem(indexPath, true);
      var button = CalculatorModel.buttons[indexPath.Row];
      model.Input(button);
    }

    [Export("collectionView:layout:sizeForItemAtIndexPath:")]
    public CGSize GetSizeForItem(UICollectionView collectionView, UICollectionViewLayout layout, NSIndexPath indexPath) {
      var button = CalculatorModel.buttons[indexPath.Row];
      var height = collectionView.Frame.Height / 5;
      if (button == "=") {
        return new CGSize(collectionView.Frame.Width / 2, height);
      } else {
        return new CGSize(collectionView.Frame.Width / 4, height);
      }
    }

    void ConfigureStyles() {
      var layout = (UICollectionViewFlowLayout)ButtonCollectionView.CollectionViewLayout;
      layout.MinimumInteritemSpacing = 0;
      layout.MinimumLineSpacing = 0;
      ButtonCollectionView.Layer.BorderWidth = 1.0f;
      ButtonCollectionView.Layer.BorderColor = UIColor.LightGray.CGColor;
      DisplayLabel.Layer.BorderColor = UIColor.LightGray.CGColor;
      DisplayLabel.Layer.BorderWidth = 1.0f;
    }

	}

  public class ButtonCell : UICollectionViewCell {

    public static NSString cellId = new NSString("ButtonCell");

    UILabel label;

    [Export ("initWithFrame:")]
    public ButtonCell(RectangleF frame) : base (frame) {
      ContentView.Layer.BorderColor = UIColor.LightGray.CGColor;
      ContentView.Layer.BorderWidth = 0.5f;

      SelectedBackgroundView = new UIView();
      SelectedBackgroundView.BackgroundColor = UIColor.LightGray;

      label = new UILabel();
      label.TextColor = UIColor.Black;
      label.TextAlignment = UITextAlignment.Center;
      label.Frame = new CGRect(10,10,30,30);
      ContentView.AddSubview(label);
    }

    public void SetTitle(string title) {
      label.Text = title;
    }
  }
}
